### ISSUE TITLE
IATI Data

### ISSUE THREAD

**Issue by akmiller01** [devinit/ddw-analyst-ui#202]
Current work is here:
https://github.com/akmiller01/IATI-Covid/blob/master/transaction_spec.py
https://github.com/akmiller01/IATI-Covid/blob/master/ts_main.py

We need to take `ts_main.py` which presently expects a folder of XML files as input and writes a folder of CSV files as output, and adapt it so it:

1. Takes a IATI Registry CKAN URL as input
2. Checks to see whether that CKAN URL has updated since last run (ignored on first run)
3. If it has updated, downloads the file and drops all related rows from final table with `iati_identifier` equal to those in the newly updated file
4. Appends rows from the newly updated file to the final table

**Comment by bill-anderson** [comment]
I would split this in three so that, if necessary, it could run off three independent sessions: the first polling the registry, the second the downloaded files, the third dealing with deletes (a publisher may delete a previously published activity)

**Session One** does 1 and 2 **and** downloads changed file.
**Session Two** does 3 (bar the download) and 4
**Session Three** needs some thought. One option is to store the filename and iati-identifier in 4. In subsequent loads 3 drops all rows with iati-identifiers that were logged in the previous load. This doesn't cover the scenarion where a publisher permanently deletes a file. Do we need a houskeeping **Session Four** where we compare all iati-identifiers for a publisher in DDW against all in the registry?

We also need to think about where the logs are stored and how they can be accessed.

And how does the system handle user queries being run at the same time that a file is being updated (replaced)?

(NB also that in future we will want to build a forward-looking budget table. And a sub-national locations table?)

**Comment by k8hughes** [comment]
@akmiller01 can you lead on this in the first instance?

**Comment by k8hughes** [comment]
I think we need much better documentation of this, @akmiller01 can you start by documenting it all really clear, and share something before writing code?  This might break down into more than one issue so we need to plan this really well...

**Comment by akmiller01** [comment]
Documenting how the IATI Registry API works:

1. Grab organisations from this endpoint: https://iatiregistry.org/api/3/action/organization_list
2. Take orgs from step 1 and add them to the package_search endpoint, e.g.: https://iatiregistry.org/api/3/action/package_search?q=organization:dipr&rows=1000000
3. Resultant JSON from step 2 will have key "result" which will have key "results" which will have key "resources" which will have keys "hash" "id" and "url." From the hash we should be able to tell whether the data is new, and the URL is where to directly download each piece of data.

**Comment by akmiller01** [comment]
Additionally, here's how to do a package-search by last modified:

Here's everything updated today:

https://iatiregistry.org/api/3/action/package_search?fq=metadata_modified:[2020-06-30T00:00:00Z%20TO%20NOW]&rows=1000000

Or everything updated in the last 24 hours:

https://iatiregistry.org/api/3/action/package_search?fq=metadata_modified:[NOW-1DAY/DAY%20TO%20NOW]&rows=1000000

**Comment by wakibi** [comment]
@akmiller01 thanks for those pointers.

**Comment by akmiller01** [comment]
@bill-anderson I've just tested out a proof of concept IATI Registry loader. Here is the logic in plain English:

1. Check for the timestamp of when the loader was last run. If no timestamp is found, return January 1, 0000 as the timestamp.
2. Query the IATI registry for all datasets updated since the last run. This can optionally given an offset (1 hour or 1 day before last run) if we're worried about missing activities that were added while the loader was running.
3. For each new dataset, generate transaction specification flat data.
4. Calculate all unique iati_identifiers contained in new data.
5. Drop all transaction rows in the "iati_transactions" table which match the iati_identifiers in the new file.
6. Append the flat data to the "iati_transactions" table.

Running my proof of concept locally for all files updated since yesterday detected approximately 6,300 activities and 160,465 transactions and took exactly 26 minutes and 17 seconds to flatten, drop duplicates, and append.

**Comment by bill-anderson** [comment]
Excellent! 
Which means we can run this on a nightly job - say 04:00 GMT - using a one hour offset?
Have you split this into separate routines? 1 and 2 should be independent of 3.
Thinking to the future is there a need to separate 3. into a dataset handler that does common jobs which then calls transaction job, budget job, etc.
You should also be storing the activity xml blob in the database.
And where are the logs written?

And we do still need a document for any future developer to comprehend (and which @wakibi should sign off on)

**Comment by akmiller01** [comment]
https://docs.google.com/document/d/1bJCXkCl8kfw5TpdeP0x0uUZOtT36xKsJM12lmmAhV8Y/edit?usp=sharing

**Comment by k8hughes** [comment]
Hi @akmiller01 how is this work going?  It would be great to get an ETA so Bill's team know when it will be ready :)

**Comment by akmiller01** [comment]
Hi Kate,

Finished on Thursday but there was some disagreement about how to store the bulk XML. Main PR here: https://github.com/devinit/ddw-analyst-ui/pull/204

And once a decision is made, the XML will either be stored in S3: https://github.com/devinit/ddw-analyst-ui/pull/223 ($5 per month extra).

Or local storage (no extra cost, but potentially slower): https://github.com/devinit/ddw-analyst-ui/pull/224

**Comment by k8hughes** [comment]
For decisions like this it might be good to have them on the issues so they are not lost.  
@wakibi  what do you think the right answer to the storage question would be?

**Comment by k8hughes** [comment]
@akmiller01 should this be in the `review` column now?

**Comment by k8hughes** [comment]
@akmiller01  - I've just spoken to Bill and he has said to go ahead with the S3 option - which sounds fine to me.

**Comment by k8hughes** [comment]
@akmiller01 @edwinmp  I am assuming that we can move this issue to `done` - are there any PR's it should have been linked to?

**Comment by akmiller01** [comment]
This one needs review and merging @k8hughes https://github.com/devinit/ddw-analyst-ui/pull/247

**Comment by k8hughes** [comment]
@akmiller01 this is still marked as in progress but I think it might have been done now as #247 has been merged - could you update the status of the issue?
